Hw02 unpack string

mrvin opened the pull request:
## Домашнее задание №2 «Распаковка строки»

### Критерии оценки
- [x] Пайплайн зелёный - 4 балла
- [x] Добавлены новые юнит-тесты - до 4 баллов
- [x] Понятность и чистота кода - до 2 баллов

#### Зачёт от 7 баллов

mrvin committed:
hw02: unpacking of repeated characters/runes

mrvin committed:
hw02: add new test case

mrvin committed:
hw02: fix module path

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,107 @@
+	for _, ch := range str {
+		switch state {
+		case start:
+			startState(&state, ch)
```
Обычно обработчик стейта принимает текущий стейт и возвращает новый: `state = startState(state, ch)`. В таком случае получаются методы без побочных эффектов.

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,107 @@
+	switch {
+	case unicode.IsDigit(ch):
+		num := int(ch - '0')
+		if _, err := resultStr.WriteString(strings.Repeat(string(prevCh), num)); err != nil {
```
`Write` методы у стрингбилдера никогда не возвращают ошибку.

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,107 @@
-func Unpack(_ string) (string, error) {
-	// Place your code here
-	return "", nil
+func digitState(state *state, ch rune) {
```
digitState полностью дублирует startState и это неспроста. Входящую строку можно разбить на несколько подстрок, каждую из которых можно обработать этой стейт-машиной. `a4bc2d5e` -> `a4`+`bc2`+`d5`+`e`. Предлагаю избавиться от стейта digit.

mrvin replied:
@kulti 
Спасибо за ревью кода. Замечания исправил. Проверьте, пожалуйста.

kulti reviewed (dismissed):
Хорошее понятно решение через стейт-машину 👍 

Код получился местами грязноватый. Если хотите 10 баллов, то предложения по улучшению кода в дискашенах.

Пока принимаю ДЗ на 9 из 10.

mrvin committed:
hw02: remove error handling for strings.Builder.Write()

Because the strings.Builder.Write...() methods always return
error = nil.

mrvin committed:
hw02: state handlers return new state

Methods should not have side effects.

mrvin committed:
hw02: delete state digit

Because digit state duplicates start state.

mrvin committed:
hw02: add function SplitAfterDigit(str string) []string

It splits the string into digit-terminated substrings, not counting
the escaped digits.

mrvin committed:
hw02: disable linter 'exhaustive'

Warns about missing case in switch of type state: 'exit'

mrvin committed:
hw02: add a loop for each substring

mrvin committed:
hw02: fix indents

mrvin committed:
hw02: add new test case

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,122 @@
+
+	splitStr := SplitAfterDigit(str)
+
+	for _, substr := range splitStr {
```
Не нужно разбивать строку на кусочки.

Я говорил о том, что когда алгоритм идет по строке a4bc2d5e и доходит до 4, то потом он, можно считать, начинает распаковывать новую строку bc2d5e тем же алгоритмом, начиная со стейта start.

printState так и делает. SplitAfterDigit  и этот внешний цикл не нужны.

mrvin replied:
Убрал функцю SplitAfterDigit и лишний цикл.

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,122 @@
+			prevCh = ch
+		}
+
+		if state != start {
```
А если backSlash? Добавьте, пожалуйста, тест на этот случай.

mrvin replied:
@kulti
Добавил тест и исправил ошибку. Проверьте, пожалуйста.

kulti reviewed (changes requested):
Стейт машина стала проще, но появился ненужный цикл и вкралась ошибка.

mrvin committed:
hw02: delete function SplitAfterDigit and excess loop

mrvin committed:
hw02: add new test case with one back slash

mrvin committed:
hw02: fix error with final state 'backSlash'

mrvin committed:
hw02: fix formatting

kulti commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,90 @@
+
+	state := start
+	for _, ch := range str {
+		switch state { //nolint:exhaustive // Warns about missing cases in switch of type state: exit
```
Здорово, что не просто отключили линтер, но и объяснили почему.

Вроде, можно не возвращать ошибку из *State() методов и добавить обработку стейта exit. Но и без этого уже все отлично.

mrvin replied:
@kulti
Исправил.

kulti reviewed (dismissed):
Куд улучшен, ошибки починены. Поднял баллы до 10.

mrvin committed:
hw02: delete error return from state handling functions
